cmd/util/ledger/migrations/staged_contracts_migration.go

```diff
@@ -148,11 +153,131 @@ func (m *StagedContractsMigration) InitMigration(
 }
 
+// collectAndRegisterStagedContractsFromPayloads scan through the payloads and collects the contracts
+// staged through the `MigrationContractStaging` contract.
+func (m *StagedContractsMigration) collectAndRegisterStagedContractsFromPayloads(allPayloads []*ledger.Payload) error {
+
+	// If the contracts are already passed as an input to the migration
+	// then no need to scan the storage.
+	// TODO: Maybe do this anyway?
+	if len(m.contractsByLocation) > 0 {
+		return nil
+	}
+
+	var stagingAccount string
+
+	switch m.chainID {
+	case flow.Testnet:
+		stagingAccount = "0x2ceae959ed1a7e7a"
+	case flow.Mainnet:
+		stagingAccount = "0x56100d46aa9b0212"
+	default:
+		// For other networks such as emulator etc. no need to scan for staged contracts.
+		return nil
+	}
+
+	stagingAccountAddress := common.Address(flow.HexToAddress(stagingAccount))
+
+	mr, err := NewMigratorRuntime(
+		stagingAccountAddress,
+		allPayloads,
+		util.RuntimeInterfaceConfig{},
+	)
+	if err != nil {
+		return err
+	}
+
+	inter := mr.Interpreter
+	locationRange := interpreter.EmptyLocationRange
+
+	storageMap := mr.Storage.GetStorageMap(stagingAccountAddress, common.PathDomainStorage.Identifier(), false)
+	iterator := storageMap.Iterator(inter)
+
+	stagedContractCapsuleStaticType := interpreter.NewCompositeStaticTypeComputeTypeID(
+		nil,
+		common.AddressLocation{
+			Name:    "MigrationContractStaging",
+			Address: stagingAccountAddress,
+		},
+		"Capsule",
+	)
+
+	stagedContracts := make([]StagedContract, 0)
+
+	for key, value := iterator.Next(); key != nil; key, value = iterator.Next() {
+		stringAtreeValue, ok := key.(interpreter.StringAtreeValue)
+		if !ok {
+			continue
+		}
+
+		storagePath := string(stringAtreeValue)
+
+		// Only consider paths that starts with "MigrationContractStagingCapsule_".
+		if !strings.HasPrefix(storagePath, "MigrationContractStagingCapsule_") {
+			continue
+		}
+
+		staticType := value.StaticType(inter)
+		if !staticType.Equal(stagedContractCapsuleStaticType) {
+			// This shouldn't occur, but technically possible.
+			// e.g: accidentally storing other values under the same storage path pattern.
+			// So skip such values. We are not interested in those.
+			continue
+		}
+
+		stagedContractCapsule := value.(*interpreter.CompositeValue)
+
+		// The stored value should take the form of:
+		//
+		// resource Capsule {
+		//     let update: ContractUpdate
+		// }
+		//
+		// struct ContractUpdate {
+		//     let address: Address
+		//     let name: String
+		//     var code: String
+		//     var lastUpdated: UFix64
+		// }
+
+		contractUpdate := stagedContractCapsule.GetField(
+			inter,
+			locationRange,
+			"update",
+		).(*interpreter.CompositeValue)
+
+		address := contractUpdate.GetField(inter, locationRange, "address").(interpreter.AddressValue)
+		name := contractUpdate.GetField(inter, locationRange, "name").(*interpreter.StringValue)
+		code := contractUpdate.GetField(inter, locationRange, "code").(*interpreter.StringValue)
```

Comment:
Maybe make this a function, and handle cast failures with errors.

Reply:
added in https://github.com/onflow/flow-go/pull/5659/commits/22c51a59f479bed1a63a29f4b58392d1b5231b57